### PR TITLE
Http download method

### PR DESCRIPTION
**Pull request by greg-1-anderson**
This PR allows the user to choose between `http` and `drush` download methods.  The use-case for http is when the user wishes to use a custom set of scaffolding files not available via drush dl, e.g. pantheon-systems/drops-8.

It is my intention that the `Extract` class should eventually become part of Robo, but I have included it here for now so that this functionality can be implemented.

### CODE REVIEW

**Comment by greg-1-anderson** [comment]
Also included in this PR is a fix that allows the scaffold update to work even if the scaffold files do not exist in the project.  I plan to submit a PR drupal-composer/drupal-project to do just that once  https://github.com/drupal-composer/drupal-project/pull/77 is merged.

**Comment by greg-1-anderson** [comment]
c.f. https://github.com/Codegyre/Robo/pull/244

**Comment by derhasi** [comment]
Would it make sense to completely switch to http and spare the drush dependency?

**Comment by greg-1-anderson** [comment]
@derhasi: I was actually thinking the same thing myself.  I'll update this PR later today.

**Comment by greg-1-anderson** [comment]
Okay, now supports http method only.  As another optimization, I made the entire scaffold operation run from a single call to rsync, by utilizing the --include filter option.  Was the rsync --delete option there for any particular reason?  Seemed like this could be too destructive, and I couldn't think of any specific reason why we would want it, so I took it out.

**Comment by webflo** [comment]
Does it mean we can get rid of the drush dependency?

**Comment by greg-1-anderson** [comment]
True; removed Drush from composer.json.

**Comment by greg-1-anderson** [comment]
Now, 'composer install' will not download the scaffold files unless there are no scaffold files on disk.  So, after the initial `composer project-create`, the scaffold files will be downloaded; however, on a CI build, if you have committed your scaffold files to the repository, then they will not be re-downloaded.

Scaffold files are always downloaded whenever `composer update` modifies `drupal/core`.

**Comment by derhasi** [comment]
I will review that today. Maybe we can split the "HTTP method" parts from the "composer install challenges" and solve those in a separate PR after discussing in #9.

**Comment by webflo** [comment]
Lets merge this one first. The scaffold is not used, this makes it easier to iterate on it. We don't have to get every right in one PR.

**Comment by derhasi** [comment]
@greg-1-anderson , due to #10, you'd need to resolve the conflicts here. :grin:

**Comment by greg-1-anderson** [comment]
Yes, I will merge in #10 shortly.

**Comment by greg-1-anderson** [comment]
This now works with the provided unit tests.  I guess the next step is to merge here, and then continue with the discussion at #9.

**Comment by greg-1-anderson** [comment]
I added one more commit here.  `sites/default/*` must also be excluded in the rsync operation.  We explicitly include `sites` and `sites/default`, because if we do not, then we cannot include anything in the `sites` or `sites/default` folders. To complement these filters, we must also exclude both `sites/*` and `sites/default/*`, so that anything not explicitly included in these locations is excluded.

**Comment by greg-1-anderson** [comment]
I misinterpreted a [comment from yesterday](https://github.com/drupal-composer/drupal-scaffold/pull/6#issuecomment-162452142) to mean "separate the discussion", but I think that "separate the implementation" was what was actually meant. I therefore removed the logic controlling the scaffold download, and made this PR invariantly download on every install and update command.

I left a vestigial `checkAction()` method, for inserting the eventual logic checks pending a satisfactory resolution at #9.

File: README.md [hunk 1]
```diff
@@ -17,15 +17,23 @@ of your root `composer.json`.
...
+      "includes": [
+        "sites/default/example.settings.my.php"
+      ],
       "omit-defaults": false
     }
   }
 }
 ```
+The `source` option may be used to specify the URL to download the
+scaffold files from; the default source is drupal.org.  The literal string
```
Comment (webflo) [inline README.md]: One space too much. ``drupal.org.  The literal string``

File: README.md [hunk 2]
```diff
@@ -41,16 +49,39 @@ example.gitignore
...
-sites
 themes
 profiles
 modules
+sites/*
+```
+
+If there are some files inside of an excluded location that should be
+copied over, they can be individually selected for inclusion via the
+`includes` option.  Default includes are provided by the plugin:
```
Comment (webflo) [inline README.md]: One space to much. ``option.  Default``

File: README.md [hunk 3]
```diff
@@ -41,16 +49,39 @@ example.gitignore
...
+sites/default/default.settings.php
+sites/default/default.services.yml
+sites/development.services.yml
+sites/example.settings.local.php
+sites/example.sites.php
+```
+
+When setting `omit-defaults` to `true`, neither the default excludes nor the
+default includes will be provided; in this instance, only those files explicitly
+listed in the `excludes` and `includes` options will be considered.  If
```
Comment (webflo) [inline README.md]: Again :)

File: src/Extract.php [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+    protected $filename;
+    protected $to;
+
+    public function __construct($filename)
+    {
+        $this->filename = $filename;
+    }
+
+  function run() {
```
Comment (webflo) [inline src/Extract.php]: The indentation looks wrong ...

File: src/Extract.php [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    // If we are not on Windows, then try to do "tar" in a single operation.
+    else {
+      $tar_compression_flag = '';
+      if ($mimetype == 'application/x-gzip') {
+        $tar_compression_flag = 'z';
+      }
+      elseif ($mimetype == 'application/x-bzip2') {
+        $tar_compression_flag = 'j';
+      }
+      $command = "tar -C $extractLocation -x{$tar_compression_flag}f {$this->filename}";
```
Comment (webflo) [inline src/Extract.php]: Does this work on Windows?

Comment (greg-1-anderson) [inline src/Extract.php]: No.  We could use the archive_tar library.

File: src/Handler.php [hunk 6]
```diff
@@ -165,6 +200,8 @@ protected function getOptions() {
     $options = $extra['drupal-scaffold'] + [
       'omit-defaults' => FALSE,
       'excludes' => [],
+      'includes' => [],
+      'source' => 'http://ftp.drupal.org/files/projects/drupal-{version}.tar.gz',
```
Comment (webflo) [inline src/Handler.php]: ftp.drupal.org supports https. :sparkles:

File: src/Extract.php (hunk 5 above)
Comment (webflo) [inline src/Extract.php]: Or composers tar downloader? https://github.com/composer/composer/blob/master/src/Composer/Downloader/TarDownloader.php#L28

Comment (greg-1-anderson) [inline src/Extract.php]: I am reluctant to use TarDownloader, because I want to move Extract.php to codegyre/Robo.

Comment (greg-1-anderson) [inline src/Extract.php]: c.f. https://github.com/Codegyre/Robo/pull/244

File: README.md (hunk 3 above)
Comment (greg-1-anderson) [inline README.md]: I guess you can tell that I learned to type on an actual typewriter.

File: src/Extract.php (hunk 5 above)
Comment (webflo) [inline src/Extract.php]: Ok, lets keep what we have.

**Comment by greg-1-anderson** [comment]
Cool; thanks.

**Pull request merged**